tests/e2e/Services/Sites/SitesCustomServerTest.php

```diff
@@ -2393,4 +2393,41 @@ public function testInvalidSSRSource(): void
 
         $this->cleanupSite($siteId);
     }
+
+    public function testPermanentRedirect(): void
+    {
+        $siteId = $this->setupSite([
+            'siteId' => ID::unique(),
+            'name' => 'Sub project site',
+            'framework' => 'other',
+            'buildRuntime' => 'node-22',
+            'outputDirectory' => './'
+        ]);
+        $this->assertNotEmpty($siteId);
+
+        $site = $this->getSite($siteId);
+        $this->assertEquals('200', $site['headers']['status-code']);
+        $this->assertArrayHasKey('adapter', $site['body']);
+
+        $domain = $this->setupSiteDomain($siteId);
+        $this->assertNotEmpty($domain);
+
+        $deploymentId = $this->setupDeployment($siteId, [
+            'code' => $this->packageSite('subProjects'),
+            'activate' => 'true'
+        ]);
+        $this->assertNotEmpty($deploymentId);
+
+        $site = $this->getSite($siteId);
+
+        $proxyClient = new Client();
+        $proxyClient->setEndpoint('http://' . $domain);
+        $response = $proxyClient->call(Client::METHOD_GET, '/');
+        $this->assertEquals(200, $response['headers']['status-code']);
+        $response1 = $proxyClient->call(Client::METHOD_GET, '/project1');
+        $this->assertEquals(200, $response1['headers']['status-code']);
+        $response2 = $proxyClient->call(Client::METHOD_GET, '/project1/');
+        $this->assertEquals(200, $response2['headers']['status-code']);
```

Comment:
Let's also ensure that body includes proper content, `Sub-directory index`, `Sub-directory project1`